onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlSTFT.h

```diff
@@ -0,0 +1,610 @@
+
+            this->hasWindowTensor = true;
+        }
+
+        // input 3: frame_length (optional; constant; scalar)
+        if (context->IsInputValid(DmlSTFTKernelInputIndex::FrameLength))
+        {
+            ComPtr<IMLOperatorTensor> tensorInterface;
+            ORT_THROW_IF_FAILED(context->GetConstantInputTensor(DmlSTFTKernelInputIndex::FrameLength, &tensorInterface));
+
+            MLOperatorTensor tensor(tensorInterface.Get());
+            uint32_t frameLength = onnxruntime::narrow<uint32_t>(OperatorHelper::ReadScalarTensorCastToInt64(tensor));
+
+            ML_CHECK_VALID_ARGUMENT(
+                this->frameSize == 0 || this->frameSize == frameLength,
+                "The window size and frame_length must be equal, if both are provided."
+            );
+
+            this->frameSize = frameLength;
+        }
+
+        ML_CHECK_VALID_ARGUMENT(this->frameSize > 0, "Either the window or frame_length must be set.");
+
+        // This limitation is a result of GpuDFTOperator in DmlDFT.h only implementing the stockham path.
+        // Remove this check if GpuDFTOperator is updated to support BluesteinZChirp.
+        ML_CHECK_VALID_ARGUMENT(
+            DFTHelpers::IsPowerOfTwo(this->frameSize),
+            "DML STFT only supports power-of-two window/frame sizes."
+        );
+
+        this->frameCount = (this->signalSize - this->frameSize) / this->frameStep + 1;
+        this->frameDftElementCount = this->isOnesided ? this->frameSize / 2 + 1 : this->frameSize;
+    }
+};
+
+using DmlSTFTShapeInferenceParams = DmlSTFTParameters<IMLOperatorShapeInferenceContextPrivate, IMLOperatorShapeInferenceContextPrivate>;
+using DmlSTFTConstructionParams = DmlSTFTParameters<IMLOperatorKernelCreationContextPrivate, IMLOperatorTensorShapeDescription>;
+
+namespace DmlSTFTHelpers
+{
+    ComPtr<ID3D12Resource> GetResourceFromKernelContext(IMLOperatorKernelContext* context, uint32_t index, bool isInput)
+    {
+        ComPtr<IMLOperatorTensor> tensor;
+        if (isInput)
+        {
+            ORT_THROW_IF_FAILED(context->GetInputTensor(index, &tensor));
+        }
+        else
+        {
+            ORT_THROW_IF_FAILED(context->GetOutputTensor(index, &tensor));
+        }
+
+        ComPtr<IUnknown> dataInterface;
+        tensor->GetDataInterface(&dataInterface);
+
+        ComPtr<ID3D12Resource> resource;
+        ORT_THROW_IF_FAILED(dataInterface.As(&resource));
+
+        return resource;
+    }
+
+    ComPtr<ID3D12Resource> GetInputResourceFromKernelContext(IMLOperatorKernelContext* context, uint32_t index)
+    {
+        return GetResourceFromKernelContext(context, index, true);
+    }
+
+    ComPtr<ID3D12Resource> GetOutputResourceFromKernelContext(IMLOperatorKernelContext* context, uint32_t index)
+    {
+        return GetResourceFromKernelContext(context, index, false);
+    }
+}
+
+// Implements STFT in two steps:
+//
```

Comment:
I appreciate how you add ASCII art comments, as it makes returning to complex code easier to understand later, and for other future maintainers.